shlog.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SHLOG" "1" "June 2016" "" ""
.
.SH "NAME"
\fBshlog\fR \- Easy logging for the shell
.
.P
fewew
.
.SH "SYNOPSIS"
\fBshlog \-l debug "I\'m a debug message"\fR
.
.P
[debug] 2016\-06\-19 23:47:36 shlog\.sh:8 \- I\'m a debug message"
.
.P
\fBSOMEVAR=3 shlog \-l info \-d SOMEVAR\fR
.
.P
[info] 2016\-06\-19 23:47:36 shlog\.sh:10 \- SOMEVAR=\'3\'
.
.SH "USAGE"
.
.SH "CONFIGURATION"
.
.SS "Log levels"
\fBshlog\fR knows five levels, from lowest to highest priority:
.
.IP "\(bu" 4
trace
.
.IP "\(bu" 4
debug
.
.IP "\(bu" 4
info
.
.IP "\(bu" 4
warn
.
.IP "\(bu" 4
error
.
.IP "" 0
.
.SS "Configuration files"
\fBshlog\fR will look in three places for a configuration file with environment variables:
.
.IP "\(bu" 4
\fB/etc/default/shlogrc\fR
.
.IP "\(bu" 4
\fB$HOME/\.config/shlog/shlogrc\fR
.
.IP "\(bu" 4
\fB$PWD/\.shlogrc\fR
.
.IP "" 0
.
.P
Environment variables in any of those files will be sourced in that order to configure the logging\.
.
.SS "Colors"
Defines the colors to use for various part of the log message if \fI\fBSHLOG_USE_STYLES\fR\fR is set for that \fIoutput\fR\.
.
.P
\fBSHLOG_STYLE_TRACE\fR : \fB%level\fR==\fBtrace\fR : Default: magenta
.
.P
\fBSHLOG_STYLE_DEBUG\fR : \fB%level\fR==\fBdebug\fR : Default: cyan
.
.P
\fBSHLOG_STYLE_INFO\fR : \fB%level\fR==\fBinfo\fR : Default: cyan
.
.P
\fBSHLOG_STYLE_WARN\fR : \fB%level\fR==\fBwarn\fR : Default: yellow
.
.P
\fBSHLOG_STYLE_ERROR\fR : \fB%level\fR==\fBerror\fR : Default: bold red
.
.P
\fBSHLOG_STYLE_MODULE\fR : \fB%module\fR : Default: bold
.
.P
\fBSHLOG_STYLE_LINE\fR : \fB%line\fR : Default: bold green
.
.P
\fBSHLOG_STYLE_DATE\fR : \fB%date\fR : Default: none
.
.P
\fBSHLOG_STYLE_RESET\fR : (after every style) : Default: \fB[0m\fR
.
.SS "Outputs"
Defines the minimum level at which to log to different outputs
.
.P
\fBSHLOG_TERM\fR: Minimum level for terminal\. Default: \fBtrace\fR
.
.P
\fBSHLOG_FILE\fR: Minimum level for file logging\. Default: \fBoff\fR
.
.SS "Enabling / Disabling colors"
Defines which outputs should use \fIstyles\fR\.
.
.P
\fBSHLOG_TERM_COLORIZE\fR: Use styles on terminal \. Default: \fB"true"\fR
.
.P
\fBSHLOG_FILE_COLORIZE\fR: Use styles on file \. Default: \fB"false"\fR
.
.SS "Term Output options"
.
.P
Whether to output to \fBstderr\fR (the default) or \fBstdout\fR\.
.
.SS "File Output options"
.
.P
\fBSHLOG_FILE_FILENAME\fR: Filename of the file to log to\.
.
.P
Default: basename of \fB$0\fR with out extension + \fB\.log\fR in \fB/tmp\fR
.
.SS "Formatting Log output"
.
.P
\fBSHLOG_DATE_FORMAT\fR: \fBstrftime(3)\fR pattern for the \fB%date\fR part of a log message, to be passed to \fBprintf\fR\.
.
.P
Default: \fB%(%F %T)T\fR (i\.e\. \fBYYYY\-MM\-DD HH:MM:SS\fR)
.
.P
\fBSHLOG_PROFILE_PRECISION\fR: The granularity with which \fI\fBshlog::profile\fR\fR will print the elapsed time\. Can be \fBns\fR for nano\-second precision or \fBms\fR for millisecond precision (the default)\.
.
.P
\fBSHLOG_FORMAT\fR: \fBprintf\fR\-Pattern for the log message\.
.
.P
Default: \fB[%level] %date %module:%line \- %msg\fR
.
.P
Custom patterns:
.
.P
\fB%level\fR: The \fIlog level\fR
.
.P
\fB%date\fR: The timestamp of the log message, see \fI\fBSHLOG_DATE_FORMAT\fR\fR
.
.P
\fB%module\fR: The calling \fImodule\fR
.
.P
\fB%line\fR: Line number of the calling script
.
.P
\fB%msg\fR: The actual log message
.
.SS "Debugging shlog"
.
.P
\fBSHLOG_SELFDEBUG\fR: If set to \fB"true"\fR, shlog will output its configuration upon first initialization\.
.
.P
Default: false
.
.SH "FAQ"
.
.SS "How to turn off logging?"
Set the \fBSHLOG_SILENT\fR variable to a non\-zero value to discard all log messages within that shell:
.
.IP "" 4
.
.nf

SHLOG_SILENT=true some\-noisy\-script\.sh
.
.fi
.
.IP "" 0
.
.SS "How to debug the logging?"
To debug the logging process itself, set the \fBSHLOG_SELFDEBUG\fR variable to a non zero value:
.
.IP "" 4
.
.nf

SHLOG_SELFDEBUG=true some\-command\.sh
.
.fi
.
.IP "" 0
.
.P
This will make \fBshlog\fR output its configuration upon initialization and also log all the files it sourced\.
.
.SS "How to log to STDOUT instead of STDERR?"
Set \fBSHLOG_TERM_OUTPUT\fR to \fBstdout\fR:
.
.IP "" 4
.
.nf

SHLOG_TERM_OUTPUT=stdout
SHLOG_TERM=debug # or trace, info, warn, error
.
.fi
.
.IP "" 0
.
.SS "How to log to a file?"
Enable the \fBfile\fR output in your \fIconfiguration file\fR:
.
.IP "" 4
.
.nf

SHLOG_FILE=trace
SHLOG_FILE_FILENAME=$PWD/myscrip\.log
.
.fi
.
.IP "" 0
.
.P
\fBSHLOG_FILE_FILENAME\fR is optional, will default to a file in /tmp derived from $0 if not set explicitly
.
.SS "How to enable or disable color output?"
Define \fBSHLOG_<output>_COLORIZE\fR\. The default is:
.
.IP "" 4
.
.nf

SHLOG_TERM_COLORIZE=true
SHLOG_FILE_COLORIZE=false
.
.fi
.
.IP "" 0
.
.SS "Logging is slow and <code>module</code> is always <code>shlog</code>?"
While you can use \fBshlog\fR as a command line script, it\'s much faster to use it as a shell function\.
.
.P
Make sure you \fBsource\fR the \fBshlog\fR script, otherwise every log command is spawning a new shell\. Compare:
.
.P
Slow:
.
.IP "" 4
.
.nf

time for i in $(seq 1000); do shlog \-l info test; done
# \.\.\.
# real  0m4\.466s
# user  0m0\.164s
# sys   0m0\.172s
.
.fi
.
.IP "" 0
.
.P
Fast:
.
.IP "" 4
.
.nf

source "$(which shlog)"
time for i in $(seq 1000); do shlog \-l info test; done
# \.\.\.
# real  0m0\.889s
# user  0m0\.408s
# sys   0m0\.136s
.
.fi
.
.IP "" 0
.
.SS "<code>shlog</code> doesn\'t respect my configuration?"
.
.SS "How to reload the configuration?"
Call \fBshlog::reload\fR, without any arguments\. It will reconfigure all variables and all changes should be visible then\.
.
.SS "How can I execute code just before the log message is printed?"
You can create a function \fBshlog::before\fR that will be called whenever \fBshlog\fR is called\. Some use cases:
.
.IP "\(bu" 4
Move the cursor to an appropriate position for full\-screen terminal apps
.
.IP "\(bu" 4
Rotate a log file
.
.IP "\(bu" 4
Add one\-off solutions for special cases (e\.g\. send an email for specific condition)
.
.IP "" 0
.
.SH "COPYRIGHT"
The MIT License (MIT)
.
.P
Copyright (c) 2016 Konstantin Baierer
.
.P
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
.
.P
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software\.
.
.P
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT\. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE\.


shlog.sh
#!/bin/bash

# BEGIN-INCLUDE LICENSE.header
# END-INCLUDE

## ## SYNOPSIS
##
##     shlog -l debug "I'm a debug message"
##     -> "[debug] 2016-06-19 23:47:36 shlog.sh:8 - I'm a debug message"
##     SOMEVAR=3
##     shlog -l info -d SOMEVAR
##     -> "[info] 2016-06-19 23:47:36 shlog.sh:10 - SOMEVAR='3'"
##

CSI="\e["
CSI_RESET="${CSI}0m"

typeset -A \
    SHLOG_LEVELS \
    SHLOG_STYLES \
    SHLOG_USE_STYLES \
    SHLOG_OUTPUTS \
    SHLOG_USE_STYLES
export \
    SHLOG_LEVELS \
    SHLOG_STYLES \
    SHLOG_OUTPUTS \
    SHLOG_USE_STYLES \
    SHLOG_FILE \
    SHLOG_SELFDEBUG \
    SHLOG_INITIALIZED

shlog () {
    local level="trace"
    local msg
    local module="${0##*/}"
    while [[ "$1" = -* ]];do
        case "$1" in
            ## ## OPTIONS

            ## ### -l,--level LEVEL
            ##
            ## Log the message to this level. Default: `trace`
            ##
            -l|--level) level=$2; shift ;;
            ## ### -m,--module MODULE
            ##
            ## Mark the log message as belonging to this module. Default: basename of `$0`
            ##
            -m|--module) module="$2"; shift ;;
            ## ### -d,--dump VARNAME
            ##
            ## Dump the definition of `VARNAME` instead of a log message.
            ##
            -d|--dump) msg=$(shlog-dump "$2"); shift ;;
            ## ### -v,--verbose
            ##
            ## Dump the configuration of shlog upon initialization. See [`SHLOG_SELFDEBUG`](#shlog-selfdebug)
            ##
            -v|--verbose) SHLOG_SELFDEBUG="true"; shift ;;
        esac
        shift
    done
    msg="${msg:-$*}"
    if [[ "$SHLOG_INITIALIZED" != "true" ]];then
        shlog-init
    fi
    local output
    for output in "${!SHLOG_OUTPUTS[@]}";do
        local out="$SHLOG_FORMAT"
        # shellcheck disable=SC2004
        output_level="${SHLOG_LEVELS[${SHLOG_OUTPUTS[$output]}]}"
        msg_level="${SHLOG_LEVELS[$level]}"
        if (( output_level > msg_level ));then
            continue
        fi
        # shellcheck disable=2059
        local date=$(printf "$SHLOG_DATE_FORMAT")
        printf -v levelfmt "%-5s" "$level"
        if [[ "${SHLOG_USE_STYLES[$output]}" = "true" ]];then
            out=${out//%level/${SHLOG_STYLES[$level]}${levelfmt}${CSI_RESET}}
            out=${out//%module/${SHLOG_STYLES[module]}${module}${CSI_RESET}}
            out=${out//%line/${SHLOG_STYLES[line]}${BASH_LINENO[0]}${CSI_RESET}}
            out=${out//%date/${SHLOG_STYLES[date]}${date}${CSI_RESET}}
        else
            out=${out//%level/$levelfmt}
            out=${out//%module/$module}
            out=${out//%line/${BASH_LINENO[0]}}
            out=${out//%date/$date}
        fi
        out=${out//%msg/$msg}
        # shellcheck disable=2059
        case "$output" in
            stdout) echo -e "$out" ;;
            stderr) echo -e "$out" >&2 ;;
            file) echo -e "$out" >> "$SHLOG_FILE" ;;
        esac
    done
}

shlog-dump () {
    local varname
    for varname in "$@";do
        {
            declare -p "$varname" 2>/dev/null|| echo "$varname="
        } |grep -o "$varname.*" #|sed -e 's,\([( ]\)\[,\1\n\t[,g'
    done
}

shlog-init () {

    ## ## ENVIRONMENT VARIABLES
    ##

    ## ### SHLOG_LEVELS
    ## 
    ## `shlog` knows five levels, from lowest to highest priority:
    ##
    ## * trace
    ## * debug
    ## * info
    ## * warn
    ## * error
    ##
    SHLOG_LEVELS=(
        [off]=999
        [trace]=0
        [debug]=1
        [info]=2
        [warn]=3
        [error]=4
    )

    ## ### SHLOG_STYLES
    ##
    ## Defines the colors to use for various part of the log message if
    ## [`SHLOG_USE_STYLES`](#shlog_use_styles) is set for that [output](#shlog-outputs).
    ##
    ## #### SHLOG_STYLE_TRACE
    ##
    ## `SHLOG_STYLE_TRACE`  : `%level` (`trace`) . Default: magenta
    ##
    ## #### SHLOG_STYLE_DEBUG
    ##
    ## `SHLOG_STYLE_DEBUG`  : `%level` (`debug`) . Default: cyan
    ##
    ## #### SHLOG_STYLE_INFO
    ##
    ## `SHLOG_STYLE_INFO`   : `%level` (`info`)  . Default: cyan
    ##
    ## #### SHLOG_STYLE_WARN
    ##
    ## `SHLOG_STYLE_WARN`   : `%level` (`warn`)  . Default: yellow
    ##
    ## #### SHLOG_STYLE_ERROR
    ##
    ## `SHLOG_STYLE_ERROR`  : `%level` (`error`) . Default: bold red
    ##
    ## #### SHLOG_STYLE_MODULE
    ##
    ## `SHLOG_STYLE_MODULE` : `%module`          . Default: bold
    ##
    ## #### SHLOG_STYLE_LINE
    ##
    ## `SHLOG_STYLE_LINE`   : `%line`            . Default: bold green
    ##
    ## #### SHLOG_STYLE_DATE
    ##
    ## `SHLOG_STYLE_DATE`   : `%date`            . Default: none
    ##
    SHLOG_STYLES=(
        [trace]="${SHLOG_STYLE_TRACE:-${CSI}35m}"
        [debug]="${SHLOG_STYLE_DEBUG:-${CSI}36m}"
        [info]="${SHLOG_STYLE_INFO:-${CSI}32;1m}"
        [warn]="${SHLOG_STYLE_WARN:-${CSI}33m}"
        [error]="${SHLOG_STYLE_ERROR:-${CSI}31;1m}"
        [module]="${SHLOG_STYLE_MODULE:-${CSI}1m}"
        [line]="${SHLOG_STYLE_LINE:-${CSI}32;1m}"
        [date]="${SHLOG_STYLE_DATE:-}"
    )

    ## ### SHLOG_OUTPUTS
    ##
    ## Defines the minimum level at which to log to different outputs
    ##
    ## #### SHLOG_OUTPUT_STDOUT
    ##
    ## `SHLOG_OUTPUT_STDOUT`: Minimum level for STDOUT. Default: `off`
    ##
    ## #### SHLOG_OUTPUT_STDERR
    ##
    ## `SHLOG_OUTPUT_STDERR`: Minimum level for STDERR. Default: `trace`
    ##
    ## #### SHLOG_OUTPUT_FILE
    ##
    ## `SHLOG_OUTPUT_FILE`: Minimum level for file logging. Default: `off`
    ##
    SHLOG_OUTPUTS=(
        [stdout]="${SHLOG_OUTPUT_STDOUT:-off}"
        [stderr]="${SHLOG_OUTPUT_STDERR:-trace}"
        [file]="${SHLOG_OUTPUT_FILE:-off}"
    )

    ## ### SHLOG_USE_STYLES
    ##
    ## Defines which outputs should use [styles](#shlog_styles).
    ##
    ## #### SHLOG_OUTPUT_STDOUT
    ##
    ##  `SHLOG_OUTPUT_STDOUT`     : Use styles on stdout . Default: `"true"`
    ##
    ## #### SHLOG_USE_STYLES_STDERR
    ##
    ##  `SHLOG_USE_STYLES_STDERR` : Use styles on stderr . Default: `"true"`
    ##
    ## #### SHLOG_USE_STYLES_FILE
    ##
    ##  `SHLOG_USE_STYLES_FILE`   : Use styles on file   . Default: `"false"`
    ##
    SHLOG_USE_STYLES=(
        [stdout]="${SHLOG_USE_STYLES_STDOUT:-true}"
        [stderr]="${SHLOG_USE_STYLES_STDOUT:-true}"
        [file]="${SHLOG_USE_STYLES_STDOUT:-false}"
    )

    ## ### SHLOG_FILE
    ##
    ## Filename of the file to log to, if logging to a file is enabled
    ##
    ## Default: basename of `0` with out extension + `.log` in `/tmp`
    ##
    SHLOG_FILE=${SHLOG_FILE:-/tmp/$(basename "$0"|sed 's/\..*$//').log}

    ## ### SHLOG_DATE_FORMAT
    ##
    ## `strftime(3)` pattern for the `%date` part of a log message, to be
    ## passed to `printf`.
    ##
    ## Default: `%(%F %T)T` (i.e. `YYYY-MM-DD HH:MM:SS`)
    ##
    SHLOG_DATE_FORMAT="${SHLOG_DATE_FORMAT:-%(%F %T)T}"

    ## ### SHLOG_FORMAT
    ##
    ## `printf`-Pattern for the log message.
    ##
    ## Custom patterns:
    ##
    ## #### %level
    ##
    ## `%level`: The [log level](#shlog-levels)
    ##
    ## #### %date
    ##
    ## `%date`: The timestamp of the log message, see [`SHLOG_DATE_FORMAT`](#shlog_date_format)
    ##
    ## #### %module
    ##
    ## `%module`: The calling [module](#-m---module-module)
    ##
    ## #### %line
    ##
    ## `%line`: Line number of the calling script
    ##
    ## #### %msg
    ##
    ## `%msg`: The actual log message
    ##
    ## Default: `[%level] %date %module:%line - %msg`
    ##
    SHLOG_FORMAT="${SHLOG_FORMAT:-[%level] %date %module:%line - %msg}"

    ## ### SHLOG_SELFDEBUG
    ##
    ## If set to `"true"`, shlog will output its configuration upon first initialization.
    ##
    ## Default: false
    ##
    if [[ "$SHLOG_SELFDEBUG" = "true" ]];then
        >&2 echo "Initialized shlog:"
        >&2 shlog-dump SHLOG_LEVELS \
            SHLOG_STYLES \
            SHLOG_USE_STYLES \
            SHLOG_OUTPUTS \
            SHLOG_FILE \
            SHLOG_FORMAT \
            SHLOG_DATE_FORMAT
    fi
    SHLOG_INITIALIZED=true
}

# detect if being sourced and
# export if so else execute
# main function with args
if [[ ${BASH_SOURCE[0]} != $0 ]]; then
  export -f shlog shlog-dump
else
  shlog "${@}"
fi
